daemon/health.go

```diff
@@ -0,0 +1,310 @@
+package daemon
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/docker/container"
+	"github.com/docker/docker/daemon/exec"
+	"github.com/docker/engine-api/types/strslice"
+)
+
+const (
+	// Longest healthcheck probe output message to store. Longer messages will be truncated.
+	maxOutputLen = 255
+
+	// Default interval between probe runs (from the end of the first to the start of the second).
+	// Also the time before the first probe.
+	defaultProbeInterval = 30 * time.Second
+
+	// The maximum length of time a single probe run should take. If the probe takes longer
+	// than this, the check is considered to have failed.
+	defaultProbeTimeout = 30 * time.Second
+
+	// Probes started within this period do not cause transitions from Starting to Unhealthy.
+	// Instead, we remain in Starting in this case (however, we can still move to Unhealthy after
+	// a successful probe, even if we're still in the grace period).
+	defaultGracePeriod = 30 * time.Second
+
+	// Shut down a container if it becomes Unhealthy.
+	defaultExitOnUnhealthy = true
+)
+
+// probeResult is used to pass the results of a probe run back to the monitor thread.
+type probeResult struct {
+	endTime  time.Time
+	exitCode int
+	output   string
+}
+
+// probe implementations know how to run a particular type of probe.
+type probe interface {
+	// Perform one run of the check. Returns the exit code and an optional
+	// short diagnostic string.
+	run(context.Context, *Daemon, *container.Container) (int, string)
+}
+
+// cmdProbe implements the "CMD" probe type.
+type cmdProbe struct{}
+
+// exec the healthcheck command in the container.
+// Returns the exit code and error message (if any)
+func (*cmdProbe) run(ctx context.Context, d *Daemon, container *container.Container) (int, string) {
```

Comment:
The error here is normally the error string returned by the probe rather than a Go error, although there are a few places where I handle the (unlikely) case of a Go error starting the probe by pretending that the error came from the probe itself. If we return these errors as a third return value, we'll just turn it into a string at the next opportunity anyway.